ksmt-core/src/main/kotlin/org/ksmt/expr/rewrite/simplify/KFpExprSimplifier.kt

```diff
@@ -0,0 +1,696 @@
+        }
+        return false
+    }
+
+    override fun <T : KFpSort> transform(expr: KFpAbsExpr<T>): KExpr<T> = simplifyApp(expr) { (arg) ->
+        if (arg is KFpValue<T>) {
+            // (abs NaN) ==> Nan
+            if (arg.isNan()) {
+                return@simplifyApp arg
+            }
+
+            if (arg.signBit) {
+                // (abs x), x < 0 ==> -x
+                val negated = arg.fpUnaryMinus()
+                negated?.let { return@simplifyApp it.uncheckedCast() }
+            } else {
+                // (abs x), x >= 0 ==> x
+                return@simplifyApp arg
+            }
+        }
+        mkFpAbsExpr(arg)
+    }
+
+    override fun <T : KFpSort> transform(expr: KFpNegationExpr<T>): KExpr<T> = simplifyApp(expr) { (arg) ->
+        if (arg is KFpValue<T>) {
+            arg.fpUnaryMinus()?.let { return@simplifyApp it.uncheckedCast() }
+        }
+
+        // (- -x) ==> x
+        if (arg is KFpNegationExpr<T>) {
+            return@simplifyApp arg.value
+        }
+
+        mkFpNegationExpr(arg)
+    }
+
+    override fun <T : KFpSort> transform(expr: KFpAddExpr<T>): KExpr<T> = expr.simplifyFpBinaryOp { rm, lhs, rhs ->
+        if (lhs is KFpValue<T> && rhs is KFpValue<T> && rm is KFpRoundingModeExpr) {
+            val result = fpAdd(rm.value, lhs, rhs)
+            result?.let { return@simplifyFpBinaryOp it.uncheckedCast() }
+        }
+
+        mkFpAddExpr(rm, lhs, rhs)
+    }
+
+    // a - b ==> a + (-b)
+    @Suppress("UNCHECKED_CAST")
+    override fun <T : KFpSort> transform(expr: KFpSubExpr<T>): KExpr<T> =
+        simplifyApp(
+            expr = expr as KApp<T, KExpr<KSort>>,
+            preprocess = { mkFpAddExpr(expr.roundingMode, expr.arg0, mkFpNegationExpr(expr.arg1)) }
+        ) {
+            error("Always preprocessed")
+        }
+
+    override fun <T : KFpSort> transform(expr: KFpMulExpr<T>): KExpr<T> = expr.simplifyFpBinaryOp { rm, lhs, rhs ->
+        if (lhs is KFpValue<T> && rhs is KFpValue<T> && rm is KFpRoundingModeExpr) {
+            val result = fpMul(rm.value, lhs, rhs)
+            result?.let { return@simplifyFpBinaryOp it.uncheckedCast() }
+        }
+
+        mkFpMulExpr(rm, lhs, rhs)
+    }
+
+    override fun <T : KFpSort> transform(expr: KFpDivExpr<T>): KExpr<T> = expr.simplifyFpBinaryOp { rm, lhs, rhs ->
+        if (lhs is KFpValue<T> && rhs is KFpValue<T> && rm is KFpRoundingModeExpr) {
+            val result = fpDiv(rm.value, lhs, rhs)
+            result?.let { return@simplifyFpBinaryOp it.uncheckedCast() }
+        }
+
+        mkFpDivExpr(rm, lhs, rhs)
+    }
+
+    @Suppress("ComplexCondition")
+    override fun <T : KFpSort> transform(expr: KFpFusedMulAddExpr<T>): KExpr<T> =
+        expr.simplifyFpTernaryOp { rm, a0, a1, a2 ->
+            if (rm is KFpRoundingModeExpr && a0 is KFpValue<T> && a1 is KFpValue<T> && a2 is KFpValue<T>) {
+                // todo: evaluate
+                mkFpFusedMulAddExpr(rm, a0, a1, a2)
```

Comment:
Is it forgotten or it's ok?

Reply:
it'is supposed to be `todo` since we have no `fma` in jdk8.